admin/fabfile.py

```diff
@@ -65,32 +66,146 @@ def bootstrap(ctx):
+    ctx.sudo("apt install -y certbot software-properties-common")
+    sudo_put(ctx, "certbot.ini", "/srv/zam/certbot.ini")
+    sudo_put(ctx, "ssl-renew", "/etc/cron.weekly/ssl-renew")
+    ctx.sudo("chmod +x /etc/cron.weekly/ssl-renew")
+    ctx.sudo(
+        "certbot certonly -c /srv/zam/certbot.ini --non-interactive " "--agree-tos"
+    )
 
 
 @task
 def sshkeys(ctx):
-    for name, key in ctx.config.get('ssh_keys', {}).items():
-        ctx.run('grep -q -r "{key}" .ssh/authorized_keys '
-                '|| echo "{key}" '
-                '| sudo tee --append .ssh/authorized_keys'.format(key=key))
+    for name, key in ctx.config.get("ssh_keys", {}).items():
+        ctx.run(
+            'grep -q -r "{key}" .ssh/authorized_keys '
+            '|| echo "{key}" '
+            "| sudo tee --append .ssh/authorized_keys".format(key=key)
+        )
 
 
 @task
 def deploy(ctx, source):
-    put_dir(ctx, quote(source), '/srv/zam/', chown='zam')
+    put_dir(ctx, quote(source), "/srv/zam/", chown="zam")
+
+
+@task
+def deploy_repondeur(
+    ctx, secret, an_pattern_liste, an_pattern_amendement, branch="master"
+):
+    user = "repondeur"
+    create_user(ctx, name=user, home_dir="/srv/repondeur")
+    clone_repo(
+        ctx,
+        repo="https://github.com/betagouv/zam.git",
+        branch=branch,
+        path="/srv/repondeur/src",
+        user=user,
+    )
+    app_dir = "/srv/repondeur/src/repondeur"
+    install_requirements(ctx, app_dir=app_dir, user=user)
+    setup_config(ctx, app_dir=app_dir, user=user, secret=secret)
+    initialize_db(ctx, app_dir=app_dir, user=user)
+    fetch_an_group_data(ctx, user=user)
+    setup_service(ctx, an_pattern_liste, an_pattern_amendement)
+
+
+@task
+def create_user(ctx, name, home_dir):
+    if ctx.sudo(f"getent passwd {name}", warn=True, hide=True).failed:
+        ctx.sudo(f"useradd --system --create-home --home-dir {quote(home_dir)} {name}")
+
+
+@task
+def clone_repo(ctx, repo, branch, path, user):
+    if ctx.run(f"test -d {quote(path)}", warn=True, hide=True).failed:
+        ctx.sudo(f"git clone --branch={branch} {repo} {quote(path)}", user=user)
+    else:
+        git = f"git --work-tree={quote(path)} --git-dir={quote(path + '/.git')}"
+        ctx.sudo(f"{git} fetch", user=user)
+        ctx.sudo(f"{git} checkout {branch}", user=user)
+        ctx.sudo(f"{git} reset --hard origin/{branch}", user=user)
+
+
+@task
+def install_requirements(ctx, app_dir, user):
+    ctx.sudo("apt-get update")
+    ctx.sudo("apt-get install --yes python3-pip")
+    ctx.sudo("python3 -m pip install pipenv")
+    ctx.sudo(f'bash -c "cd {app_dir} && pipenv install"', user=user)
+
+
+@task
+def setup_config(ctx, app_dir, user, secret):
+    template_local_file(
+        "../repondeur/production.ini.template",
+        "../repondeur/production.ini",
+        {"secret": secret},
+    )
+    sudo_put(
+        ctx, "../repondeur/production.ini", f"{app_dir}/production.ini", chown=user
+    )
```

Comment:
Then delete the local `production.ini` generated file?

Reply:
Good idea! I turned `template_local_file` into a context manager that will automatically remove the generated file after use → c7052df9609a613b43f73977a57f5a89c02a405b